### PR TITLE
(WIP) call sites

### PR DESCRIPTION
**Pull request by jasonLaster**
Opening an early PR, which we can collaborate on.

### CODE REVIEW

File: src/components/Editor/index.js [hunk 1]
```diff
@@ -693,9 +702,9 @@ class Editor extends PureComponent {
           breakpoint: bp,
           editor: this.editor && this.editor.codeMirror
         })
-      );
+      );*/
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: we should remove the column breakpoints code here and the component

File: src/components/Editor/index.js [hunk 2]
```diff
@@ -792,6 +801,15 @@ class Editor extends PureComponent {
     );
   }
 
+  renderCallSites() {
+    const editor = this.editor;
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: lets add the column bps feature flag here

File: src/components/Editor/index.js [hunk 3]
```diff
@@ -295,12 +300,30 @@ class Editor extends PureComponent {
...
+  onKeyUp(e) {
+    let { key } = e;
+
+    if (key === "Alt") {
+      e.stopPropagation();
+      e.preventDefault();
+
+      this.setState({ showCallSites: false });
+    }
+  }
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: lets move `onKeyUp` and `onKeyDown` to CallSites because we can :)

File: src/components/Editor/index.js [hunk 4]
```diff
@@ -295,12 +300,30 @@ class Editor extends PureComponent {
     }
   }
 
+  onKeyUp(e) {
+    let { key } = e;
+
+    if (key === "Alt") {
+      e.stopPropagation();
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: others might handle this too so perhaps we dont want to stop propogation

File: src/components/Editor/Editor.css [hunk 5]
```diff
@@ -5,6 +5,10 @@
   --editor-second-searchbar-height: 27px;
 }
 
+.call-sites {
+  outline: 1px solid #aed3ef;
+}
```
Comment (jasonLaster) [inline src/components/Editor/Editor.css]: this should be in its own call-sites css file

File: src/components/Editor/CallSites.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+export default connect(
+  state => {
+    const selectedLocation = getSelectedLocation(state);
+    const selectedSource = getSelectedSource(state);
+    const sourceId = selectedLocation && selectedLocation.sourceId;
+    const source = selectedSource && selectedSource.toJS();
+
+    getCallSiteBreakpoint = location => {
```
Comment (jasonLaster) [inline src/components/Editor/CallSites.js]: this should be in its own helper

File: src/components/Editor/CallSites.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    let sites;
+    if (!callSites) {
+      return null;
+    }
+
+    callSites = callSites.filter(
+      callSite => callSite.location.start.line === callSite.location.end.line
+    );
+
+    editor.codeMirror.operation(() => {
```
Comment (jasonLaster) [inline src/components/Editor/CallSites.js]: i'm not sure if this helped... lets look at the perf here. Maybe not a blocker, but nice

File: src/utils/parser/getSymbols.js [hunk 8]
```diff
@@ -115,6 +116,19 @@ export default function getSymbols(source: SourceText): SymbolDeclarations {
         });
       }
 
+      if (t.isCallExpression(path)) {
+        const callee = path.node.callee;
+        const calleeNode = t.isMemberExpression(callee)
```
Comment (jasonLaster) [inline src/utils/parser/getSymbols.js]: we should exclude member expressions cause... well you know

File: src/components/Editor/CallSites.js (hunk 7 above)
Comment (bomsy) [inline src/components/Editor/CallSites.js]: i tot it did when we tested it, at the allhands ... i'll try it again

File: src/components/Editor/CallSite.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      if (nextProps.showCallSite) {
+        if (!this.marker) {
+          this.addCallSite(nextProps);
+        }
+      } else {
+        if (!nextProps.breakpoint) {
+          this.clearCallSite();
+        }
+      }
+      // console.log(nextProps.showCallSite);
```
Comment (bomsy) [inline src/components/Editor/CallSite.js]: remove console.log

File: src/utils/parser/tests/__snapshots__/getSymbols.js.snap [hunk 10]
```diff
@@ -56,6 +56,30 @@ variables
...
 `;
 
+exports[`Parser.getSymbols call sites 1`] = `
+"properties
+
+member expressions
+[(2, 8), (2, 9)] [(2, 0), (2, 9)]  c 
+[(2, 4), (2, 5)] [(2, 0), (2, 5)]  b 
+call expressions
+   undefined
```
Comment (jasonLaster) [inline src/utils/parser/tests/__snapshots__/getSymbols.js.snap]: it looks like we might be getting the member expressions and not the call expressions with this change?

Comment (bomsy) [inline src/utils/parser/tests/__snapshots__/getSymbols.js.snap]: ah .. yep!

File: src/components/Editor/CallSites.js [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+      sites = dom.div(
+        {},
+        callSites.map((callSite, index) => {
+          return CallSite({
+            key: index,
+            callSite,
+            editor,
+            breakpoint: callSite.breakpoint,
```
Comment (bomsy) [inline src/components/Editor/CallSites.js]: we might not need to pull this in on its own since we have it on `callSite`

**Comment by jasonLaster** [comment]
I made some progress on this today, but ultimately could not get it all the way there:

done:
1. styling in light/dark theme
2. bottom border uses `::before` elements so they don't overlay
3. callsites are associated with breakpoints

not done:
1. removing a bp fails because we can't find the callsite for a location...

**Comment by bomsy** [comment]
looking into the failure on removing breakpoints might have an idea what the issue is ...

**Pull request merged**